libs/utils/include/utils/RangeMap.h

```diff
@@ -0,0 +1,309 @@
+
+    // Private helper that assumes there is no existing range that overlaps the given range.
+    void insert(KeyType first, KeyType last, const ValueType& value) noexcept {
+        assert_invariant(!has(first));
+        assert_invariant(!has(last - 1));
+
+        // Check if there is an adjacent range to the left than can be extended.
+        KeyType previous = first;
+        if (iterator iter = find(--previous); iter != end() && get_value(iter) == value) {
+            get_range(iter).last = last;
+            merge_right(iter);
+            return;
+        }
+
+        // Check if there is an adjacent range to the right than can be extended.
+        if (iterator iter = find(last); iter != end() && get_value(iter) == value) {
+            get_range(iter).first = first;
+            return;
+        }
+
+        mMap[first] = {Range<KeyType> { first, last }, value};
+    }
+
+    // Private helper that erases all intervals that are wholly contained within the given range.
+    // Note that this is quite different from the public "clear" method.
+    void wipe(KeyType first, KeyType last) noexcept {
+        // Find the first range whose beginning is greater than or equal to "first".
+        iterator iter = mMap.lower_bound(first);
+        while (iter != end() && get_range(iter).first < last) {
+            KeyType existing_last = get_range(iter).last;
+            if (existing_last > last) {
+                break;
+            }
+            iter = mMap.erase(iter);
+        }
+    }
+
+    // Check if there is range to the right that touches the given range.
+    // If so, erase it, extend the given range rightwards, and return true.
+    bool merge_right(iterator iter) {
+        iterator next = iter;
+        if (++next == end() || get_value(next) != get_value(iter)) {
+            return false;
+        }
+        if (get_range(next).first != get_range(iter).last) {
+            return false;
+        }
+        get_range(iter).last = get_range(next).last;
+        mMap.erase(next);
+        return true;
+    }
+
+    // Check if there is range to left right that touches the given range.
+    // If so, erase it, extend the given range leftwards, and return true.
+    bool merge_left(iterator iter) {
+        iterator prev = iter;
+        if (--prev == end() || get_value(prev) != get_value(iter)) {
+            return false;
+        }
+        if (get_range(prev).last != get_range(iter).first) {
+            return false;
+        }
+        get_range(iter).first = get_range(prev).first;
+        mMap.erase(prev);
+        return true;
+    }
+
+    // Erase the given range if it contains no elements.
+    void clean(iterator iter) {
+        Range<KeyType>& range = get_range(iter);
+        assert_invariant(range.first <= range.last);
+        if (range.first == range.last) {
+            mMap.erase(iter);
+        }
+    }
+
+    const_iterator find(KeyType key) const noexcept {
```

Comment:
can one `find` be implemented in term of the other one, e.g. using `const_cast`?  It seems strange to have the same exact code duplicated twice.

Reply:
Fixed.  Tried a few different ways (including `const_cast` and `std::as_const`) but settled on using a template helper method because that's the easiest way to handle both `const_iterator` and `iterator` return types.